Add guide for ServiceChain upgrade & hard fork (#467)
klaytn/klaytn-docs: docs/node/service-chain/references/hard-fork.md

Klaytn has introduced several [hard forks](../../klaytn-history) to provide new features.

> In accordance with the Klaytn hard forks, new versions of ServiceChain binaries are also released to support the hard forks. However you need to activate the hard fork in order to use their new features in the ServiceChain.

Author: I thought that the reasons to release ServiceChain binaries v1.9.0 were:
1. Make ServiceChain binaries compatible with their parent chain (Klaytn v1.9.0)
2. ServiceChain can activate Magma fork too

So the v1.9.0 release of ServiceChain supports Magma hard fork on both parent chain and ServiceChain itself.

And for the second sentence, I tried to emphasize that upgrading ServiceChain to v1.9.0 and activating the Magma hard fork are two distinct operations. So I would prefer to update that statement if it seems ambiguous instead of removing it.

I will try to elaborate more on these in our meeting.

> In accordance with the Klaytn hard forks, new versions of ServiceChain binaries are also released to support the hard forks. However you need to activate the hard fork in order to use their new features in the ServiceChain.

Author: Closing because this is outdated

> **NOTE** Upgrading ServiceChain binaries may be irreversible and backward-incompatible, meaning you cannot downgrade to an older version.

Reviewer: May I ask why you think the `scn` binary change is irreversible and has backward incompatibility? For example, parent(v1.8.4) <---> child(v1.9.0) works, and parent(v1.8.4) <---> child(v1.8.4) works too.

Author: What I meant was that once you upgrade `scn` to v1.9.0, you cannot go back to v1.8.4.
```
Fatal: Error starting protocol stack: database version is v4, Klaytn v1.8.4 only supports v3
```

Also, I think `parent(v1.8.4) <---> child(v1.9.0)` or `parent(v1.8.4) <---> child(v1.8.4)` cases are not relevant to above content. Anyway, `parent(v1.8.4) <---> child(v1.9.0)` didn't work when I tested.

Reviewer: I had a misunderstanding about it. Okay good. I think pasting the quote for the database compatibility issue from the release note is better for more correctness. 

The not working case of `parent(v1.8.4) <---> child(v1.9.0)` is due to the incompatible protocol version. I misread it. Sorry.

Author: OK, I will add the below comment copied from v1.9.0 release note as an example.

```bash
$ kscnd stop
Shutting down kscnd: OK
$ cp /path/to/new/kscn /path/to/original/kscn
```

**NOTE** After you have replaced all ServiceChain nodes' binaries, keep the ServiceChain nodes down (or stopped). You will restart them after you have activated the hard fork.

## 3. Activate Hard Fork <a id="3-activate-hard-fork"></a>

If you have upgraded the ServiceChain binaries with a version that supports desired hard fork, you can activate the hard fork in the ServiceChain by re-initializing the chain config with updated genesis.

### Update genesis and re-initialize chain config for all ServiceChain nodes <a id="update-genesis-and-re-initialize-chain-config-for-all-servicechain-nodes"></a>

First, specify the hard fork number in the `config` field of `genesis.json`.
For example, if you are trying to activate the Magma hard fork in your ServiceChan, you should specify the `magmaCompatibleBlock` in `config` field of the genesis, like below.

```json
{
  "config": {
    "chainId": 1000,
    "istanbulCompatibleBlock": 0,
    ...
    "magmaCompatibleBlock": 1500,
    ...
  },
  ...
}
```

> To enable a hard fork in the chain config, the previous hard fork should be enabled.
> That is, to enable Magma hard fork, the EthTxType hard fork should be already enabled.
> If there are missing hard forks in the chain config, you have to add them too.

Reviewer: ```suggestion
To enable a hard fork in the chain config, the previous hard forks should be enabled.
That is, to enable Magma hard fork, the EthTxType hard fork should be already enabled.
If there are missing fields for the compatible block numbers of preceding hard forks hard forks in the chain config, you have to add them too.
```

Author: Done, thanks!

```bash
$ kscn attach ~/kscnd_home/klay.ipc
Welcome to the Klaytn JavaScript console!

instance: Klaytn/vX.X.X/XXXX-XXXX/goX.X.X
 datadir: ~/kscnd_home
 modules: admin:1.0 debug:1.0 eth:1.0 governance:1.0 istanbul:1.0 klay:1.0 net:1.0 personal:1.0 rpc:1.0 txpool:1.0 web3:1.0

> governance.vote("kip71.lowerboundbasefee", 0)
"Your vote is prepared. It will be put into the block header or applied when your node generates a block as a proposer. Note that your vote may be duplicate."
> governance.vote("kip71.upperboundbasefee", 0)
"Your vote is prepared. It will be put into the block header or applied when your node generates a block as a proposer. Note that your vote may be duplicate."
```

If the votes for updating the upper and lower bound of the gas price were successful, those changes will take effect after 2 istanbul epochs (An epoch has the value in block numbers).

For example, if the epoch is 3600, and the votes for updating the upper and lower bounds of gas price has been placed in the block #4000, those changes will take effect starting from the block #10800.

> In detail, the votes will be finalized when their first epoch was encounted in the block #7200, and the changes are applied when their second epoch was encountered in the block #10800.

Reviewer: ```suggestion
In detail, the votes will be finalized when their first epoch reaches block #7200, and the changes are applied at the second epoch (#10800).
```

Author: I slightly modified your suggestions, please take another look and see if you have further comments.
